wl/server/ServerUtils.java

```diff
@@ -1,15 +1,103 @@
+/*
+ * Copyright (C) 2021 by the Widelands Development Team
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ */
+
 package wl.server;
 
 import java.io.*;
 import java.net.*;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
 import java.util.*;
+import java.util.concurrent.*;
 import wl.utils.*;
 
 abstract class ServerUtils {
 	public static final ThreadActivityAndGitHubSyncManager SYNCER =
 	    new ThreadActivityAndGitHubSyncManager();
 	public static final Random RANDOM = new Random(System.currentTimeMillis());
 
+	public static void log(String msg) {
+		System.out.println("[" + new Date() + " @ " + Thread.currentThread().getName() + "] " +
+		                   msg);
+	}
+
+	public static interface Functor { public void run() throws Exception; }
+
+	private static final Map<String, Semaphore> _semaphores = new HashMap<>();
+	private static final int SEMAPHORE_BLOCK_RW_ACCESS = 100;
+	private static synchronized Semaphore getSemaphore(String addon) {
+		if (!_semaphores.containsKey(addon))
+			_semaphores.put(addon, new Semaphore(SEMAPHORE_BLOCK_RW_ACCESS, true));
+		return _semaphores.get(addon);
+	}
+
+	public static void semaphoreRO(String addon, Functor f) throws Exception {
+		doSemaphore(addon, 1, f);
+	}
+	public static void semaphoreRW(String addon, Functor f) throws Exception {
+		doSemaphore(addon, SEMAPHORE_BLOCK_RW_ACCESS, f);
+	}
+	private static void doSemaphore(String addon, final int i, Functor f) throws Exception {
+		Exception throwme = null;
+		Semaphore s = getSemaphore(addon);
+		s.acquire(i);
+		try {
+			f.run();
+		} catch (Exception e) {
+			throwme = e;
+		}
+		s.release(i);
+		if (throwme != null) throw throwme;
```

Comment:
This looks like a self-built `finally` block. Try:
```suggestion
		} catch (Exception e) {
			throw e;
		} finally {
		    s.release(i);
		}
```
See https://docs.oracle.com/javase/specs/jls/se16/html/jls-14.html#jls-14.20.2

Reply:
Thanks, changed it